File: frontend/src/components/benchmark/content/quick_sort/implementations/quick_sort.java
import java.util.*;

public class QuickSort {
    private int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    private int partitionRandom(int[] arr, int low, int high) {
        Random rand = new Random();
        int randomIdx = rand.nextInt(high - low + 1) + low;
        swap(arr, randomIdx, high);
        return partition(arr, low, high);
    }

    private void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private void quickSortRecursive(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSortRecursive(arr, low, pi - 1);
            quickSortRecursive(arr, pi + 1, high);
        }
    }

    private void quickSortRandomRecursive(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partitionRandom(arr, low, high);
            quickSortRandomRecursive(arr, low, pi - 1);
            quickSortRandomRecursive(arr, pi + 1, high);
        }
    }

    public void sort(int[] arr) {
        quickSortRecursive(arr, 0, arr.length - 1);
    }

    public void sortRandom(int[] arr) {
        quickSortRandomRecursive(arr, 0, arr.length - 1);
    }

    public void sortIterative(int[] arr) {
        Stack<int[]> stack = new Stack<>();
        stack.push(new int[]{0, arr.length - 1});

        while (!stack.isEmpty()) {
            int[] range = stack.pop();
            int low = range[0];
            int high = range[1];

            if (low < high) {
                int pi = partition(arr, low, high);
                stack.push(new int[]{low, pi - 1});
                stack.push(new int[]{pi + 1, high});
            }
        }
    }

    private void threeWayPartition(int[] arr, int low, int high) {
        if (low >= high) return;

        int lt = low, gt = high;
        int pivot = arr[low];
        int i = low;

        while (i <= gt) {
            if (arr[i] < pivot) {
                swap(arr, lt, i);
                lt++;
                i++;
            } else if (arr[i] > pivot) {
                swap(arr, i, gt);
                gt--;
            } else {
                i++;
            }
        }

        threeWayPartition(arr, low, lt - 1);
        threeWayPartition(arr, gt + 1, high);
    }

    public void sortThreeWay(int[] arr) {
        threeWayPartition(arr, 0, arr.length - 1);
    }
}

File: frontend/src/components/benchmark/content/quick_sort/implementations/quick_sort.js
class QuickSort {
    partition(arr, low, high) {
        const pivot = arr[high];
        let i = low - 1;

        for (let j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
        [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
        return i + 1;
    }

    partitionRandom(arr, low, high) {
        const randomIdx = Math.floor(Math.random() * (high - low + 1)) + low;
        [arr[randomIdx], arr[high]] = [arr[high], arr[randomIdx]];
        return this.partition(arr, low, high);
    }

    quickSortRecursive(arr, low, high) {
        if (low < high) {
            const pi = this.partition(arr, low, high);
            this.quickSortRecursive(arr, low, pi - 1);
            this.quickSortRecursive(arr, pi + 1, high);
        }
    }

    quickSortRandomRecursive(arr, low, high) {
        if (low < high) {
            const pi = this.partitionRandom(arr, low, high);
            this.quickSortRandomRecursive(arr, low, pi - 1);
            this.quickSortRandomRecursive(arr, pi + 1, high);
        }
    }

    sort(arr) {
        this.quickSortRecursive(arr, 0, arr.length - 1);
    }

    sortRandom(arr) {
        this.quickSortRandomRecursive(arr, 0, arr.length - 1);
    }

    sortIterative(arr) {
        const stack = [[0, arr.length - 1]];

        while (stack.length > 0) {
            const [low, high] = stack.pop();

            if (low < high) {
                const pi = this.partition(arr, low, high);
                stack.push([low, pi - 1]);
                stack.push([pi + 1, high]);
            }
        }
    }

    threeWayPartition(arr, low, high) {
        if (low >= high) return;

        let lt = low, gt = high;
        const pivot = arr[low];
        let i = low;

        while (i <= gt) {
            if (arr[i] < pivot) {
                [arr[lt], arr[i]] = [arr[i], arr[lt]];
                lt++;
                i++;
            } else if (arr[i] > pivot) {
                [arr[i], arr[gt]] = [arr[gt], arr[i]];
                gt--;
            } else {
                i++;
            }
        }

        this.threeWayPartition(arr, low, lt - 1);
        this.threeWayPartition(arr, gt + 1, high);
    }

    sortThreeWay(arr) {
        this.threeWayPartition(arr, 0, arr.length - 1);
    }
}

module.exports = QuickSort;